File: core/Splits/Ids.cs
namespace Optimus.Core;

/// <summary>
/// Identity of a pane (a leaf in the split tree that owns an ordered tab list). Wraps a
/// monotonically increasing <see cref="int"/> issued by <see cref="SplitTreeController"/> and
/// <b>never reused</b> within a session (KTD6), so a stale snapshot can never alias a pane that
/// was created after the snapshot was captured (R12).
/// </summary>
public readonly record struct PaneId(int Value)
{
    public override string ToString() => $"P{Value}";
}

/// <summary>
/// Identity of a surface — one terminal, i.e. one engine instance (ConPTY + render thread),
/// presented as one tab. Monotonic and never reused within a session (KTD6).
/// </summary>
public readonly record struct SurfaceId(int Value)
{
    public override string ToString() => $"S{Value}";
}

/// <summary>
/// Identity of a split branch (an interior node). Lets the view map a divider drag back to the
/// exact branch regardless of how the tree is later re-shaped — the Windows analog of bonsplit's
/// per-split UUID (optimus/Sources/Workspace.swift <c>setDividerPosition(_:forSplit:)</c>). Stable
/// across rewrites: an operation that does not touch a branch preserves its id.
/// </summary>
public readonly record struct BranchId(int Value)
{
    public override string ToString() => $"B{Value}";
}

/// <summary>
/// Mints pane/surface/branch ids, monotonic and never reused (KTD6). Phase 5 introduces multiple
/// workspaces, each with its own <see cref="SplitTreeController"/>; sharing one allocator across all
/// of them keeps <see cref="SurfaceId"/>s globally unique, so a surface id alone (e.g. from
/// <c>OPTIMUS_SURFACE_ID</c> over the Phase-4 pipe) unambiguously identifies its workspace.
/// </summary>
public sealed class IdAllocator
{
    private int _pane;
    private int _surface;
    private int _branch;

    public PaneId NextPane() => new(++_pane);
    public SurfaceId NextSurface() => new(++_surface);
    public BranchId NextBranch() => new(++_branch);
}

File: shell/src/domain/ids.rs
//! Port of core/Splits/Ids.cs — pane/surface/branch identity, monotonic and never reused
//! within a session (KTD6), so stale snapshots can never alias later-created panes (R12).

use std::fmt;

/// Identity of a pane (a leaf in the split tree that owns an ordered tab list).
#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug, PartialOrd, Ord)]
pub struct PaneId(pub i32);

impl fmt::Display for PaneId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "P{}", self.0)
    }
}

/// Identity of a surface — one terminal (one engine instance), presented as one tab.
#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug, PartialOrd, Ord)]
pub struct SurfaceId(pub i32);

impl fmt::Display for SurfaceId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "S{}", self.0)
    }
}

/// Identity of a split branch (an interior node). Stable across rewrites: an operation
/// that does not touch a branch preserves its id.
#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug, PartialOrd, Ord)]
pub struct BranchId(pub i32);

impl fmt::Display for BranchId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "B{}", self.0)
    }
}

/// Mints pane/surface/branch ids. One allocator is shared across all workspaces so
/// `SurfaceId`s are globally unique — a surface id alone (e.g. `OPTIMUS_SURFACE_ID` over
/// the pipe) unambiguously identifies its workspace.
#[derive(Default)]
pub struct IdAllocator {
    pane: i32,
    surface: i32,
    branch: i32,
}

impl IdAllocator {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn next_pane(&mut self) -> PaneId {
        self.pane += 1;
        PaneId(self.pane)
    }

    pub fn next_surface(&mut self) -> SurfaceId {
        self.surface += 1;
        SurfaceId(self.surface)
    }

    pub fn next_branch(&mut self) -> BranchId {
        self.branch += 1;
        BranchId(self.branch)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ids_are_monotonic_and_start_at_one() {
        let mut alloc = IdAllocator::new();
        assert_eq!(alloc.next_pane(), PaneId(1));
        assert_eq!(alloc.next_pane(), PaneId(2));
        assert_eq!(alloc.next_surface(), SurfaceId(1));
        assert_eq!(alloc.next_surface(), SurfaceId(2));
        assert_eq!(alloc.next_branch(), BranchId(1));
    }

    #[test]
    fn counters_are_independent_per_kind() {
        let mut alloc = IdAllocator::new();
        alloc.next_pane();
        alloc.next_pane();
        alloc.next_pane();
        assert_eq!(alloc.next_surface(), SurfaceId(1));
        assert_eq!(alloc.next_branch(), BranchId(1));
    }

    #[test]
    fn display_matches_csharp_tostring() {
        assert_eq!(PaneId(3).to_string(), "P3");
        assert_eq!(SurfaceId(7).to_string(), "S7");
        assert_eq!(BranchId(2).to_string(), "B2");
    }
}
